modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java

```diff
@@ -849,4 +857,60 @@ private NotificationSender notificationSender(long requestId) {
         // This is fine, because the client registers a listener before sending the request.
         return (writer, err) -> sendNotification(requestId, writer, err);
     }
+
+    @Override
+    public CompletableFuture<Boolean> notify(AuthenticationEventParameters parameters, @Nullable Throwable exception) {
+        if (shouldCloseConnection(parameters)) {
+            LOG.warn("Closing connection due to authentication event [connectionId=" + connectionId + ", remoteAddress="
+                    + channelHandlerContext.channel().remoteAddress() + ", event=" + parameters.type() + ']');
+            closeConnection();
+        }
+        return falseCompletedFuture();
+    }
+
+    private boolean shouldCloseConnection(AuthenticationEventParameters parameters) {
+        readWriteLock.readLock().lock();
```

Comment:
I think it is better to get the lock in `currentUserAffected` method then `AUTHENTICATION_ENABELED` and `default` won't hold the lock they are not needed in.